### ISSUE TITLE
[Upstream Bug] VSCode + Pylance + Pyright linter errors on "beartype.typing.TypeAlias"

### ISSUE THREAD

**Issue by rbroderi** [beartype/beartype#127]
I get errors when trying to use the  beartype.typing.TypeAlias instead of typing.TypeAlias,
here are some of the error messages I get:
Key: TypeAlias = str   
Expression of type "Type[str]" cannot be assigned to declared type "TypeAlias"
  "Type[type]" is incompatible with "Type[TypeAlias]"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)

CallbackFunction: TypeAlias = Callable[[Key], Any]
Expression of type "(Unknown) -> Any" cannot be assigned to declared type "TypeAlias"
  "object" is incompatible with "TypeAlias"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)

def press(key: Key):
    print(f"'{key}' pressed")
Expected class type but received "TypeAlias"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)

**Comment by leycec** [comment]
Clearly, Pylance popped too many benzos last night.

`beartype.typing.TypeAlias` is *literally* just `typing.TypeAlias`. The former is a convenient synonym for the latter; the two are the exact same object. Specifically, this is how the `beartype.typing` subpackage declares `TypeAlias`

```python
            from typing import (  # type: ignore[attr-defined]
                ...
                TypeAlias as TypeAlias,
                ...
            )
```

**That's it.** `beartype.typing` just imports `typing.TypeAlias` into itself under the exact same name. This can't be simplified any further, because this is already maximally simple. It's just a simple import, which means that Pylance can't handle even simple imports. This is where I facepalm myself repeatedly on a Reddit livestream to farm upvotes.

## What We Goin' Do Is This

Would you mind filing an upstream issue against **[Pyright's issue tracker](https://github.com/microsoft/pyright/issues)**, @rbroderi?

You are probably now wondering why I am now talking about "Pyright," because weren't you just talking about Pylance? Right. So, Pyright is Microsoft's static type-checker. Under the hood, Pylance leverages Pyright to statically check types. The type-checking failures you've exhibited are actually being directly emitted by Pyright – not Pylance. Pylance is just forwarding Pyright complaints to you.

This is a critical issue with Pyright, which appears to be unable to cope with [PEP 484-compliant implicit type aliases](https://peps.python.org/pep-0484/#type-aliases). That's bad, because PEP 484 is nearly a decade old. Every type checker should *at least* support PEP 484.

To quote PEP 484:

> Type aliases are defined by simple variable assignments:

```python
Url = str
def retry(url: Url, retry_count: int) -> None: ...
```

> **Type aliases may be as complex as type hints** in annotations – anything that is acceptable as a type hint is acceptable in a type alias:

```python
from typing import TypeVar, Iterable, Tuple

T = TypeVar('T', int, float, complex)
Vector = Iterable[Tuple[T, T]]

def inproduct(v: Vector[T]) -> T:
    return sum(x*y for x, y in v)
```

Pyright appears to be failing at that. :facepalm: :man_facepalming: :woman_facepalming:

**Comment by leycec** [comment]
Please feel free to reopen this if there's anything tangible we can do for you, Pylance, or Pyright to make this work. I *loathe* closing issues unresolved – but this deflated ball is almost certainly in Pyright's court.

Oh, and thanks for the recent spate of issues! Drop us a line at the bear cave if you hit any further snags or come up with any h0t feature requests.

**Comment by rbroderi** [comment]
Thank you for the information and I will open an issue with Pyrights upstream issue tracker.

**Comment by leycec** [comment]
Thanks *so* much, @rbroderi. I know how time-consuming merely filing GitHub issues can be, because that's pretty much all I did last night. Monday evening, I want you back! :sweat_smile:

@beartype applauds the amazing work you're doing for us all.

**Comment by rbroderi** [comment]
opened as issue https://github.com/microsoft/pyright/issues/3363 on pyright. Documenting here for for discoverability.

**Comment by agucova** [comment]
Issue was closed upstream because Pyright only checks for `typing.TypeAlias` and won't implement deeper checking.

**Comment by leycec** [comment]
*...heh.* Knew it. Upstream's toxic, so it's little surprise they refuse to fix their blatantly broken and PEP-noncompliant behaviour. As [detailed above](https://github.com/beartype/beartype/issues/127#issuecomment-1101971619), [PEP 484 requires type checkers to support implicit type aliases](https://peps.python.org/pep-0484/#type-aliases):

> Type aliases may be as complex as type hints in annotations – **anything that is acceptable as a type hint is acceptable in a type alias.**

`typing.TypeAlias` is a type hint. So, any type alias of `typing.TypeAlias` is also acceptable as a type hint:

```python
# Pyright should support this simple type alias.
from typing import TypeAlias
MyTypeAlias = TypeAlias
my_str_type: MyTypeAlias = str

# Pyright should support this simple type alias, too.
from typing import TypeAlias as MyTypeAlias
my_str_type: MyTypeAlias = str
```

Pyright doesn't support either. Ergo, Pyright violates PEP 484 – Python's core typing standard. To no one's surprise, Pyright then refuses to fix their standards violations. :face_exhaling: 

And Pyright's excuse for violating PEP 484...

> TypingAlias is one of a handful of special forms that pyright needs to handle early in the "binding" phase, prior to type evaluation.

**...is not a valid excuse.** Mypy complies with PEP standards. Pyright needs to, too.

If Pyright currently can't because they're improperly handling `TypeAlias` "early in the 'binding' phase," then they need to stop doing that and start doing what mypy does. Ironically, Pyright is wrong; mypy is right. Low-level implementation details are *no* excuse for perpetuating wrong behaviour.

> Why was `beartype.typing.TypeAlias` created in the first place?

The `beartype.typing` submodule is a one-for-one mapping of the `typing` module. Everything that `typing` provides, `beartype.typing` also provides and vice versa.

`beartype.typing` is a drop-in replacement for `typing`. Ideally, @beartype users should globally replace *all* instances of `typing` throughout their codebases with `beartype.typing`.

This is a good thing – *not* a bad thing. `beartype.typing` is demonstrably superior to `typing` in increasingly many ways. mypy supports these ways, because mypy is PEP-compliant; Pyright violates these ways, because Pyright is PEP-noncompliant.

This is why standards matter, people. Come on, Microsoft! Throw the community a bone here.

**Comment by leycec** [comment]
**Reluctantly reopening,** because upstream refuses to resolve their broken behaviour. It's unclear whether we can circumvent this on behalf of Pylance VSCode users – *but maybe we can.*

At the least, reopening this will improve discoverability and hopefully provoke some fruitful discussion.

## tl;dr

Pylance VSCode users, please **prefer `typing.Alias` to `beartype.typing.TypeAlias`** for the moment. This isn't @beartype's fault, because we comply with PEP 484.

**Comment by leycec** [comment]
**"Clever" yet paradoxically stupid idea:** @beartype could *probably* silence this VSCode complaint by:

1. Detecting when it's being statically type-checked by Pyright.
2. If so, replacing `beartype.typing.TypeAlias` with `typing.Any`. `TypeAlias` is *mostly* semantically meaningless anyway, so just aliasing to to `Any` *mostly* does little harm. Of course, that entirely defeats the purpose of using `TypeAlias` in the first place... but false positives would still be better than what's happening now.

So, something like this is what we're saying:

```python
# In "beartype.typing.__init__":
if TYPE_CHECKING and IS_TYPE_CHECKER_PYRIGHT:
    TypeAlias = Any
else:
    from typing import TypeAlias as TypeAlias
```

Should work – maybe, in theory. But Pyright's kinda cray-cray, so maybe it'd still expurgate everywhere even if we feed it that. 

## B-b-but... How Do We Detect Pyright?

**I have no idea.** I doubt anyone else does, either. The usual approach is something like:

```python
IS_TYPE_CHECKER_PYRIGHT = 'pyright' in sys.modules
```

Of course, that's fragile and error-prone. Thanks alot, Redmond.

**Comment by leycec** [comment]
**Okay.** Just begging Pyright to "go away already" is clearly an insufficient solution, despite being my preferred solution. Let's brainstorm.

I don't :heart: VSCode, because I :heart_on_fire: Vim. But if I *did* :heart: VSCode, here's what I personally would do:

* **Disable Pylance type-checking,** because Pylance uses Pyright to statically type-check by default. Pyright demonstrably violates core Python type-checking standards like PEP 484, which is bad. Specifically:
   1. Open the `User Settings` dialog.
   2. Search for `Type Checking Mode`.
   3. Find `Python › Analysis: Type Checking Mode` in the resulting search hits.
   4. Switch the "default rule set for type checking" to `off`.
* **Enable mypy type-checking,** because mypy complies with literally all Python type-checking standards. In fact, mypy is the *de facto* Python type-checking standard against which all other type checkers (including @beartype and Pyright) are measured. Specifically:
   1. [Install **mypy**](https://mypy.readthedocs.io/en/stable/getting_started.html).
   2. [Install the **Mypy VSCode extension**](https://marketplace.visualstudio.com/items?itemName=matangover.mypy).
   3. Rejoice as things actually work for once.

Data scientists and machine learning specialists especially want to switch from Pyright to mypy, because numerous scientific Python packages of interest provide mypy but *not* Pyright plugins – because [Pyright devs refuse to implement a comparable plugin API on grounds of fundamentalist religious purity](https://github.com/microsoft/pyright/issues/607#issuecomment-873467941). :facepalm: 

## Time to Open a Can of Whoopie

Because Pylance's bad type-checking makes us look and feel bad, this is mission critical. I'm going to immediately add a new FAQ entry with this exact advice.

If someone who does :heart: VSCode could verify that the above instructions are both valid *and* make this issue go away, I'll happily close this after amending our FAQ.

Thanks for all the heads up, all. Everyone rocks! :rock:

**Comment by leycec** [comment]
**[Glorious FAQ entry that solves everything is up](https://github.com/beartype/beartype#id27).**

Since `pyright` refuses to fix its violations of typing standards required by @beartype, our only recourse is to advise all VSCoders (...*this means you*) to [embrace the VSCode Mypy extension](https://marketplace.visualstudio.com/items?itemName=matangover.mypy). Let's pretend I actually tested those instructions. :sweat_smile: 

Big thanks to @rbroderi for this and the companion issue #126!

**Comment by rsokl** [comment]
> Pyright devs seem toxic, so it's little surprise they refuse to fix their blatantly broken and PEP-noncompliant behaviour. 

You might want to pump your brakes here a bit, my friend. pyright is basically a one-man show, which is pretty incredible when you think about how many people depend on pyright via VSCode... it probably has more (inadvertent) users than any other static type checker in the python ecosystem. Moreover, Eric is an indispensable member of the typing community, and I have found him to be quite generous with his time.

> our only recourse is to advise all VSCoders (...this means you) to embrace the VSCode Mypy extension

Simply opting for the mypy VSCode extension comes with some serious downsides. Beyond pyright being much more performant than mypy speed wise (it takes mypy 60s to scan the `src/` of hydra-zen vs pyright taking <6s!), pyright also supports some important typing features that mypy lacks (which would make them non-compliant with several PEPs...). These include:
- recursive types
- [variadic generics](https://peps.python.org/pep-0646/)
- [`functools.partial`](https://github.com/python/mypy/issues/12593) (This is why `jax` has an opaque `jit` wrapper  -- it returns `Wrapped` like a coward -- because mypy can't handle it! pyright totally handles partial'd decorators without issue)
- [dataclass transforms](https://peps.python.org/pep-0681/)
- [Self](https://peps.python.org/pep-0673/)

It is absolutely bonkers how quickly (and correctly) pyright implements new typing features! For all of these reasons, I prefer pyright over mypy, [even outside of the context of VSCode](https://github.com/mit-ll-responsible-ai/hydra-zen/runs/6892249698?check_suite_focus=true). (Also, did you know that it has a really cool [tool for measuring the type completeness of a library](https://github.com/microsoft/pyright/blob/92b4028cd5fd483efcf3f1cdb8597b2d4edd8866/docs/typed-libraries.md#verifying-type-completeness)?)

I just wanted to share this because I don't think that pyright should be underestimated in its importance nor its utility. And more generally, I just want peace and love among the various Python type-checkers, static and runtime alike ☮️ 

P.S. the whole time I was writing this I kept thinking of the ad

![image](https://user-images.githubusercontent.com/29104956/173746243-efc99e84-c542-4c58-a304-7712f4cbe8b4.png)

but I liked Nintendo as a kid! So I felt like a bad guy who was peddling rude, in-your-face advertising lines from the 90s  while writing this 😞

**Comment by leycec** [comment]
*...heh.* :heart_on_fire: the SEGA trash talk. Pretty sure I heard the [SEGA startup sound](https://www.youtube.com/watch?v=9dIXw0OEtk4) while reading that.

I hear and abide by everything that you are saying. You are not wrong. In fact, you're right. Like beartype, mypy *is* deficient in various respects. This kinda makes sense, right? Eric's a Technical Fellow at Microsoft. It's his job to get Pythonic typing correct. Meanwhile, most of the rest of us in the open-source typing community are mostly unpaid open-source volunteers. It's hard to compete. We're just here for the sheer joy of type hints. <sup>*how is that even a thing*</sup>

## Ain't Nobody Got Time for This

Ultimately, my conundrum is that open-source volunteer time is scarce. I'd fervently *love* to openly embrace mypy, pyright, [Google's pytype](https://github.com/google/pytype), and every other static type checker. B-b-but... I lack the bandwidth for that. Even supporting mypy is an ongoing manly struggle with destiny, deliverance, and `# type: ignore[EVERYTHING-JUST-IGNORE-EVERYTHING-MYPY]` debacles.

Mypy is Python's official static type checker. It's clunky, it's chunky, and it's funky. But it's also community-maintained, CLI-friendly, continuous integration (CI)-friendly, pure-Python, lightweight, and the *de facto* typing standard.

Pyright is none of those things. I don't know how to properly integrate pyright into either our local `pytest` **+** `tox` test suite *or* our remote GitHub Actions-based CI testing workflow – if pyright even can be. I don't even know how to properly install pyright as a system-wide package under Gentoo Linux, because pyright is crazily implemented in TypeScript (i.e., typed JavaScript) rather than pure-Python. Pyright requires `node`, which requires `npm`. Pyright is heavyweight. Pyright requires that I throw scarce resources at non-standard solutions that I personally will never leverage, because it's Vim all the way down at the @leycec cottage.

Pyright is hostile to my interests. Mypy isn't.

Honestly? I truly abhor all static type checkers. Python isn't statically typed. It's dynamically typed. Trying to pretend otherwise is an exercise in futility and `# type: ignore[...]` spam. The *real* reason @beartype supports mypy is that I can force mypy to get out @beartype's way with basically no effort in 2022. That reasoning currently doesn't extend to pyright.

## But Reality Always Bats Last

Of course... this isn't the social reality of the situation. The social reality is, as you say, that pyright is inadvertently deployed across a large horizontal slice of Python's dev space. But the personal reality is that Microsoft isn't paying me to support pyright; they're paying Eric to support runtime type checkers like @beartype.

I want everyone to use @beartype. So, I want @beartype to support all actively maintained static type checkers. This includes pyright. If I can enable pyright to trivially support @beartype without a vast expenditure of effort on my part, I happily will. But... I don't know how to do that.

Thus, the conundrum continues. `</sigh>`

**Comment by leycec** [comment]
I have good news. I have bad news.

## Give It to Me Straight, Doc

The good news is @beartype now officially supports `pyright`! Rejoice, typing gentleman – for we have vanquished a vacuous evil this day:

```
$ pyright beartype
No configuration file found.
No pyproject.toml file found.
stubPath /home/leycec/py/beartype/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 177 source files
pyright 1.1.254
0 errors, 0 warnings, 0 informations 
Completed in 13.868sec
```

**0 errors, 0 warnings, 0 informations,** people.

<sub>Yes, `pyright` wants us to believe that "informations" is the plural form of "information." It isn't; like the few remaining hairs on my shining head, [the English noun "information" is uncountable and thus has no plural](https://jakubmarian.com/information-is-vs-informations-are-is-information-plural-or-singular-in-english). And... that should tell you everything you need to know about `pyright`.</sub>

But I digress. Ultimately, this means we have resolved the related issue #126. `pyright` should no longer say this (...or anything, really) when confronted with @beartype in real-world code:

```
Type of "beartype" is partially unknown
Type of "beartype" is "Overload[(obj: BeartypeableT@beartype) ->
    BeartypeableT@beartype, (*, conf: BeartypeConf) -> 
   ((BeartypeableT@BeartypeConfedDecorator) -> 
   BeartypeableT@BeartypeConfedDecorator),
   (obj: BeartypeableT@beartype, *, conf: BeartypeConf = BeartypeConf()) -> 
   (BeartypeableT@BeartypeReturn | ((BeartypeableT@BeartypeConfedDecorator) ->
   BeartypeableT@BeartypeConfedDecorator))]"
```

Our upcoming `beartype 0.11.0` release thus extends static type-checking support to both `mypy` *and* `pyright`. Pretty sure that makes us the first Python package to explicitly support both – but I'm pretty sure someone in the peanut gallery is about to prove me wrong. <sup>*Do it!*</sup>

Until then, please dangerously install @beartype from `git` to silence `pyright` complains:

```bash
python -m pip install git+https://github.com/beartype/beartype.git@720d779a1a4fdb830d6374cc5dc40e6e54153042
```

## So... How Bad Is this Bad News?

**Not that bad,** actually! How often does that happen, guys?

The bad news is that *this* issue that we are discussing here still isn't a @beartype issue, as expected. It's an upstream `pyright` issue. Fairly certain we can't resolve this on our end. I'd happily monkey-patch this if we could with a sketchy kludge that makes our test suite sweat bullets – but `pyright` is TypeScript, so there's nothing to monkey-patch.

Instead, we need `pyright` to do something about this. But Eric isn't inclined to do anything about this, because he's Eric. Sadly, the `upstreamwontfix` tag is more appropriate than even I suspected.

## Do Something, You! Do Something or Face Wrath!

*...yeah.* I totally get it. I'd be annoyed, too.

The only consolation here is that `beartype.typing.TypeAlias` is just a trivial alias for `typing.TypeAlias`. So, please just do this for the moment:

```python
# Instead of this...
from beartype.typing import TypeAlias

# ...please do this!
from typing import TypeAlias

# Pyright should no longer complain about this.
MyTypeHint: TypeAlias = bool | int | str | None
```

With apologies, I hang my head as tears stream down my dirt-streaked cheeks. Thus ends another tale of woe and debauchery in the pits of Python typing. :smiling_face_with_tear:

**Comment by rsokl** [comment]
>  I have bad news.

> 
```python
# Instead of this...
from beartype.typing import TypeAlias

# ...please do this!
from typing import TypeAlias
```
> With apologies...

This is the least bad bad news ever, but this just goes to show that @beartype's standards are sky high! 🚀 🚀 🚀 

This is awesome! @beartype is looking real

![image](https://user-images.githubusercontent.com/29104956/175321111-91e16ce4-f496-4778-b9a3-04dd4c82874b.png)


right now. Thank you for this incredible work and for suffering through the herculean effort of supporting both pyright and mypy.

>  Pretty sure that makes us the first large-scale Python package to explicitly support both

You may be right! At least the first large-scale Python package that is heavily steeped in Pythons typing system!

**Comment by leycec** [comment]
**OMG.** That apropos promo shot of a 128-in-1 MegaDrive cart stacked on a classic trash Data East SNES cart stacked on an obscure NES cart with a boss logo even I have never seen before but am now desperately scavenging Google Image for.

I see your 8- and 16-bit childhood nostalgia bit-packing and raise you... *the MegaDrive Tower of Doom.*

![shit just went down](https://pbs.twimg.com/media/DTmboslVwAA1Nm_?format=jpg&name=900x900)

Unrelatedly, [GitHub's OpenAI Codex-fueled Copilot](https://github.com/features/copilot) just launched. What is this nightmare fuel in IDE form? Now paired programming is an abstract cloud-hosted quasi-daemonic entity critiquing everything I code in real time against everything Guido and Soklaski are coding in real time. That's just great, guys.

Pretty sure this timeline ends with a [creepy yet wise Alan Watts AI](https://www.youtube.com/watch?v=Exe7yup2Yu0) snarkily suggesting that He will catalyze the Singularity sooner than I can release `beartype 1.0.0`. And what will gall the most is... *Alan Watts AI will not be wrong.*
